### ISSUE TITLE
[Performance] Performance problems with large numbers of metrics

### ISSUE THREAD

**Issue by mrodikov** [go-graphite/carbonapi#518]
**Problem description**
The main problem is that the query unfolds in about 65 thousand metrics, resulting in about 65 thousand find queries and some render queries to graphite-clickhouse.
Individual queries are not a problem, for example, find queries are executed in 0.1 seconds on average, render queries are executed in 0.4 seconds. The only problem is that there are several tens of thousands of queries, which causes carbonapi query to run for about 30 minutes.

Estimated at best, 65,000 requests will be executed sequentially in 108 minutes. 65000*0.1 - 6500 seconds = 108 minutes.
We have 4 CPUs, so assuming they are evenly loaded, we will execute 65,000 requests in 27 minutes, which corresponds to the real request response time.

**carbonapi's version**
0.14.0 and custom build from dev-master (for extended debug logging)

**Does this happened before**
I don't think that depends on the version

**carbonapi's config**
I've used different configurations, but the problem appears in any case.
The configuration is almost identical to the reference one, except for the backendsv2 section.
`
upstreams:

    backendsv2:
        backends:
            -
                groupName: "clickhouse"
                protocol: "auto"
                lbMethod: "broadcast"
                maxTries: 3
                maxBatchSize: 100
                keepAliveInterval: "30s"
                concurrencyLimit: 250
                maxIdleConnsPerHost: 200
                timeouts:
                    find: "5000s" #was 2s
                    render: "5000s" #was 120s
                    connect: "300ms"
                servers:
                    - "http://127.0.0.1:9090"
`

**backend software and config**
graphite-clickhouse 0.11.7, Clickhouse server 20.3.5.21

**carbonapi performance metrics**
The problem is not in the API, but in the number of requests, I do not think that the data required in this section will be useful in any way.

**Query that causes problems**
1. What's the query?
http://127.0.0.1:8081/render/target=aliasByNode(summarize(sumSeriesWithWildcards(A.B.C.*.*.*.*.*.*.*.*.*.D.E.{F,G}.count, 3, 4, 5, 6, 7, 8, 9, 10, 11), '01d', 'sum', false), 5)&from=-7d&format=json&maxDataPoints=1553

2. How many metrics does it fetch?
About 65,000.

**Additional context**
I did not create this issue because something is slower than it should be, but rather to get some advice on what to do in that case (apart from increasing the number of CPUs or decreasing the number of wildcards in the request).

**Comment by mrodikov** [comment]
The answer time decreased to 8-9 minutes. I think it will not be possible to achieve better results by changing the configuration, so the issue can be closed.

Thank you so much for your help.

**Comment by Felixoid** [comment]
hey @mrodikov, I just wonder, do you use the latest tag of https://github.com/lomik/graphite-clickhouse, or the master?

If the tag, could I ask you to measure the difference, please?

**Comment by mrodikov** [comment]
The last result was from tag. I will do the test again with the master.

**Comment by mrodikov** [comment]
> If the tag, could I ask you to build the master and measure the difference, please?

I've got the same response times for master ~8 minutes

**Comment by Civil** [comment]
If you can, could you also test https://github.com/lomik/graphite-clickhouse/pull/85 ?

It could improve performance of complex dashboards a bit.

**Comment by mrodikov** [comment]
Yes, I will try tomorrow with carbonapi@master and graphite-clickhouse@37b91df.
I guess I will also need to switch the protocol to 'carbonapi_v3_pb' in the configuration.

We only have 1 Clickhouse backend now (lbMethod is 'rr' anyway).

**Comment by mrodikov** [comment]
Unfortunately, with carbonapi_v3_pb the query ended with the error "Internal Server Error: error or no response: no metrics in the Response" after 33m.
This time the problem was not in the context timeout (find/render), but because of an error in graphite-clickhouse: "clickhouse response status 400: Code: 168, DB:Exception: AST is too big. Maximum 50000".
maxBatchSize has been set to 0, I will try with a different one today. 
maxBatchSize = 0, looks like a bad choice for carbonapi_v3_pb

**Comment by Civil** [comment]
maxBatchSize other than 0 is even worse for Clickhouse in general (checking if metric is there for huge amount of metrics would take more time than actually fetching the data).

For the query to complete you might need to tune clickhouse as well:
https://github.com/ClickHouse/ClickHouse/issues/2710

and relevant docs:
https://clickhouse.tech/docs/en/operations/settings/query-complexity/#max-ast-elements

**Comment by Civil** [comment]
Another interesting thing here - that it should have any difference in how graphite-clickhouse interacts with clickhouse itself for the query that you've provided and the only diffrence there should be - avoiding extra copies to convert format from carbonapi_v2 to v3 (carbonapi does that internally) and on grahpite-clickhouse side it should send just list of floats with metadata, instead of list of floats + list of bools for carbonapi_v2.

Only difference is that if you use timeshift function then request for both normal and time-shifted data would be send in the same request to graphite-clickhouse.

**Comment by mrodikov** [comment]
No luck with max_ast_elements=100000
graphite-clickhouse consumed like 80% of available memory (clickhouse-server workers consumed the rest) and carbonapi process ran out of memory (fatal error: runtime: out of memory)

I have 24Gb available, should have been enough.

**Comment by Civil** [comment]
Interesting, as the memory consumption with the PR and v3 pb should be actually about 10% lower.

You are still using the query from the initial post? Or the actual query is different?

**Comment by mrodikov** [comment]
I'm using the same query everytime.

**Comment by Civil** [comment]
Overall: https://golang.org/doc/diagnostics.html#profiling

and https://blog.golang.org/pprof

Or specifically for memory leaks: https://medium.com/dm03514-tech-blog/sre-debugging-simple-memory-leaks-in-go-e0a9e6d63d4d

Overall it might be better if I'll be able to reproduce your workload. But for that I would need to know more about the query and workload:
 1. You've mentioned 65k metrics matching. If I'd create 70k that would be more than you currently have? 
 2. How many datapoints per metirc graphite-clickhouse should return? (or retention schema of it?)
 3. What's the distribution of metrics matching those stars?

**Comment by mrodikov** [comment]
Metrics in this query refer to the visitor counter.

> 1. You've mentioned 65k metrics matching. If I'd create 70k that would be more than you currently have?

With 
```sql
SELECT COUNT(DISTINCT Path) FROM graphite_index WHERE Path LIKE 'A.B.C%.reserved.invariant.visitor.count'
```

 I get 64103 metrics.
With 
```sql
SELECT COUNT(DISTINCT Path) FROM graphite_index WHERE Path LIKE 'A.B.C%.reserved.invariant.visitor_mutated.count'
```

I get 554 metrics.
So it looks like 70000 will be enough.

> 2. How many datapoints per metirc graphite-clickhouse should return? (or retention schema of it?)

If I understood the question correctly, I need information about how many datapoints can be in each individual metric.
With 
```sql
SELECT t.cnt, COUNT(*) FROM (
    SELECT COUNT(*) AS cnt FROM graphite WHERE Path LIKE 'A.B.C%.reserved.invariant.visitor.count' AND Timestamp > (NOW() - toIntervalDay(7)) GROUP BY Path
) t GROUP BY cnt ORDER BY COUNT(*) DESC
```

I get
```bash
cnt COUNT  
21 5060  
42 579  
63 184  
84 92
105 50
126 31
22 27
168 22
147 21
189 11
...
5076 1
4829 1
...
13060 1
12492 1
```

It turns out that the vast majority of metrics has up to a hundred datapoints.

> 3. What's the distribution of metrics matching those stars?

If I understand correctly, I need to provide here the number of possible value variants for each wildcard in the query.
With
```sql
SELECT COUNT(DISTINCT arrayElement(splitByChar('.', Path), %n)) FROM graphite_index WHERE Path LIKE 'A.B.C%.reserved.invariant.visitor.count'
```

%n - index of a metric part
```bash
4 100
5 61
6 238
7 9043
8 111
9 41652
10 7
11 7
12 8
```

**Comment by Felixoid** [comment]
Hey @mrodikov, here's my first attempt to pass aggregation to clickhouse https://github.com/lomik/graphite-clickhouse/pull/88, so gch will respect maxDataPoints and delegate aggregation with proper steps and functions to CH.

There one necessary setting needed to enable new behavior

```
[clickhouse]
internal-aggregation = true
```

I'd really appreciate it if you could try it. As far as I could, I've caught a lot of bugs already, but it's still WIP. So, to see if PoC works and worth to try.

**Comment by Civil** [comment]
So far I can't reproduce this behavior. For me `carbonapi_v3_pb` is about 15% faster and about half of memory consumption on carbonapi side (and about same on graphite-clickhouse).

So I've used that config:
```
listen: "localhost:8081"
expvar:
  enabled: true
  pprofEnabled: true
  listen: ""
headersToPass:
  - "X-Dashboard-Id"
  - "X-Grafana-Org-Id"
  - "X-Panel-Id"
concurency: 1000
cache:
   type: "mem"
   size_mb: 0
   defaultTimeoutSec: 60
   memcachedServers:
       - "127.0.0.1:1234"
       - "127.0.0.2:1235"
cpus: 0
tz: ""
maxBatchSize: 0
graphite:
    host: ""
    interval: "60s"
    prefix: "carbon.api"
    pattern: "{prefix}.{fqdn}"
idleConnections: 1000
pidFile: ""
upstreams:
    buckets: 10
    timeouts:
        find: "120s"
        render: "600s"
        connect: "200ms"
    concurrencyLimitPerServer: 1000
    keepAliveInterval: "30s"
    maxIdleConnsPerHost: 1000
    backendsv2:
        backends:
          -
            groupName: "clickhouse-cluster1"
            protocol: "carbonapi_v3_pb"
            lbMethod: "rr"
            maxTries: 3
            maxBatchSize: 0
            keepAliveInterval: "30s"
            concurrencyLimit: 1000
            maxIdleConnsPerHost: 1000
            timeouts:
                find: "120s"
                render: "600s"
                connect: "200ms"
            servers:
                - "http://127.0.0.1:9090"
graphTemplates: /home/civil/graphite/graphTemplates.example.yaml
expireDelaySec: 10
logger:
    - logger: ""
      file: "stderr"
      level: "debug"
      encoding: "console"
      encodingTime: "iso8601"
      encodingDuration: "seconds"
    - logger: ""
      file: "carbonapi.log"
      level: "info"
      encoding: "json"

```

To reproduce the results I've used following configuration:
I've populated clickhouse with 55k metrics, all of them had random number of datapoints from 1 to 5000. Retention schema on clickhouse side was set to 1s for forever.

Query I did:
http://localhost:8081/render/?from=1597872404&until=1598477204&format=json&target=sum(i518.*.*.*.*.*.*.*.*)

As sumSeriesWithWildcard would be harder to verify and it shouldn't behave differently with different backend protocol (I've compared raw results before applying functions and it was the same).

**Comment by mrodikov** [comment]
> I've populated clickhouse with 55k metrics (70k is a bit too much for my home machine as even during this test it took about 30 GB of RAM at peak for both graphite-clickhouse and carbonapi), all of them had random number of datapoints from 1 to 5000. 

And how much RAM did the request with 55k metrics take?

I wrote about high RAM consumption, on the current versions of carbonapi (0.14.0) and graphite-clickhouse (0.11.7) I see the use of about 2.5 GB of RAM, when using graphite-clickhouse version supporting carbonapi_v3_pb, only graphite-clickhouse uses about 20 GB before carbonapi crashes. This increase in RAM consumption is a normal expected behavior of  graphite-clickhouse?

**Comment by mrodikov** [comment]
> Hey @mrodikov, here's my first attempt to pass aggregation to clickhouse [lomik/graphite-clickhouse#88](https://github.com/lomik/graphite-clickhouse/pull/88), so gch will respect maxDataPoints and delegate aggregation with proper steps and functions to CH.
> 
> There one necessary setting needed to enable new behavior
> 
> ```
> [clickhouse]
> internal-aggregation = true
> ```
> 
> I'd really appreciate it if you could try it. As far as I could, I've caught a lot of bugs already, but it's still WIP. So, to see if PoC works and worth to try.

8 minutes
I don't see in logs that queries to clickhouse change somehow (I added internal-aggregation = true in /etc/graphiteclickhouse/graphite-clickhouse.conf). How do I know that internal aggregation works?

**Comment by Civil** [comment]
> I wrote about high RAM consumption, on the current versions of carbonapi (0.14.0) and graphite-clickhouse (0.11.7) I see the use of about 2.5 GB of RAM, when using graphite-clickhouse version supporting carbonapi_v3_pb, only graphite-clickhouse uses about 20 GB before carbonapi crashes. This increase in RAM consumption is a normal expected behavior of graphite-clickhouse?

In my experiment above I've seen gch to take about 14 GB of RAM with a very small difference between carbonapi_v2 and v3 protocols.

carbonapi in case of carbonapi_v2 used about 21 GB of RAM and in case of carbonapi_v3 - about 12 GB of ram.

Anyway memory consumption shouldn't change a a lot unless you have a query that fetced very few (less than 100) datapoints in the reply and even then it should be 20 GB vs 2.5 GB.

For the protocol changes you can compare:
https://github.com/go-graphite/protocol/blob/master/carbonapi_v3_pb/carbonapi_v3_pb.proto#L46-L60 - carbonapi_v3_pb response
https://github.com/go-graphite/protocol/blob/master/carbonapi_v2_pb/carbonapi_v2_pb.proto#L7-L14 - carbonapi_v2_pb response

TLDR;
 * `isAbsent` which was array of bool values was removed in v3_pb. This should save 1 * amount_of_points bytes (it's encoded as 1 byte and `values` field still have fixed64 value on that place). Also this is not for amount of fetched datapoints, but for amount of datapoints in the response (`(stop - start) / step` to be precise)
 * v3_pb have `pathExperession` (original target), `xFilesFactor` (fixed32 - 4 bytes, always 0 in case of gch), `highPrecisionTimestamps` (1 byte),  `appliedFunctions` (currently empty), `requestStartTime` and `requestStopTime` which are original timestamps for the request. In case if something is empty it won't be passed on the wire (appliedFunctions are not for example). So total overhead for metadata would be `len(pathExpression) + len("Average") + 4 + 1 + 8 + 8 = from 28 to about 100 bytes.`

This is per-target.

**Comment by mrodikov** [comment]
Next week I will try to do an experiment with carbonapi_v3_pb again, perhaps something just went wrong this time.

**Comment by Felixoid** [comment]
> How do I know that internal aggregation works?

The query should look like `SELECT Path, anyResample(1598013900, 1598618699, 6300)(toUInt32(intDiv(Time, 6300)*6300), Time), avgOrNullResample(1598013900, 1598618699, 6300)(Value, Time)`, where the start of `avgOrNullResample` depends on the aggregate function from `system.graphite_retention`

**Comment by mrodikov** [comment]
> > How do I know that internal aggregation works?
> 
> The query should look like `SELECT Path, anyResample(1598013900, 1598618699, 6300)(toUInt32(intDiv(Time, 6300)*6300), Time), avgOrNullResample(1598013900, 1598618699, 6300)(Value, Time)`, where the start of `avgOrNullResample` depends on the aggregate function from `system.graphite_retention`

Can't see any of these.

Probably I should have changed the repository path in makefile before running "make"?

**Comment by Felixoid** [comment]
hmmm... it's enough to checkout the repository and run `make` in the root

**Comment by Civil** [comment]
@Felixoid As far as I understand, your changes and clickhouse-side aggregation currently requires to use carbonapi_v3_pb on carbonapi side and for now won't work with carbonapi_v2_pb

**Comment by Felixoid** [comment]
Not necessarily, for v2 proto it still tries to parse it from url https://github.com/Felixoid/graphite-clickhouse/blob/clickhouse_aggregation/render/handler.go#L94-L96 and falls back to the config, where default is 4096 https://github.com/Felixoid/graphite-clickhouse/blob/clickhouse_aggregation/config/config.go#L175

So with `internal-aggregation = true` it should at least change the queries https://github.com/Felixoid/graphite-clickhouse/blob/clickhouse_aggregation/config/config.go#L77

**Comment by Felixoid** [comment]
I've deployed the latest commit to our production system and it looks very promising. If you've built something on Friday or weekend, I'd recommend to use the current one, since there was a lot of improvements.

**Comment by mrodikov** [comment]
Sorry for the long absence, I will try to check the latest versions tomorrow.

**Comment by Felixoid** [comment]
Any update regarding the experiment with a later version?

**Comment by Felixoid** [comment]
I assume, it's fast enough :-j

**Comment by mrodikov** [comment]
I am sorry, I was too busy to do a comparative test. I have added time for testing to my current sprint, so the results will be during the current week.

**Comment by mrodikov** [comment]
Since we moved to another server (technically it is identical to the previous one), the software versions have changed and I made tests with the next versions of the software:
0) no load, no custom carbonapi or graphite-clickhouse running [LA <1, memory 2/24G]
1) carbonapi 0.14.0, graphite-clickhouse 0.11.7, clickhouse-server 20.4.5.36 (current) [LA <2, memory 2.5/24G] - ~1m30s
2) carbonapi dev-master, graphite-clickhouse 0.11.7 [LA <2, memory 2.5/24G] - ~1m30s
3) carbonapi dev-master, graphite-clickhouse dev-master (carbonapi_v3_pb pull request was merged into master so I took dev-master) - no luck, didn't fit into wget 900s timeout [LA >20-30, memory 24/24G, icinga LA alert]
4) carbonapi dev-master, Felixoid/graphite-clickhouse dev-master [LA <2, memory 2.5/24G]  - ~1m30s
5) carbonapi dev-master, Felixoid/graphite-clickhouse dev-master + internal-aggregation = true [LA <2, memory 2.5/24G]  - ~1m30s

Query: 
```text
http://127.0.0.1:8081/render/?target=aliasByNode(summarize(sumSeriesWithWildcards(business.prod.traffic.*.*.*.*.*.*.*.*.*.reserved.invariant.{visitor,visitor_mutated}.count), '01d', 'sum', false), 5)&from=1602115200&until=now&format=json&maxDataPoints=1571
```



(5) With internal-aggregation = true queries like "SELECT Path, groupArray(Time), groupArray(Value)..." appeared in graphite-clickhouse.log

(3) With graphite-clickhouse dev-master I see in carbonapi.log that backend name changed from "clickhouse-cluster_protobuf" to "clickhouse-cluster_carbonapi_v3_pb". Also, carbonapi & graphite-clickouse consuming ~7G of RAM each while in other test cases they're consuming like 200M each

In general, at the moment I don't have any performance problems.
I hope that this information will be useful.

**Comment by Felixoid** [comment]
Since you mention, that with (5) you see `SELECT Path, groupArray(Time), groupArray(Value)..` in logs, it's very suspicious.

Also, could you, please, test `carbonapi dev-master, graphite-clickhouse dev-master + internal-aggregation = true`?

**Comment by mrodikov** [comment]
Any attempt to use graphite-clickhouse@dev-master results in the complete utilization of the processor and RAM (Probably if I change the protocol from "auto" to v2_pb, the problem will disappear, but I didn't check it.).

So I decided to double-check the options with carbonapi@dev-master and Felixoid/graphite-clickhouse with and without internal-aggregation enabled (I may have missed something in previous checks).

(4) 
```bash
# A lot of
SELECT Path FROM graphite.graphite_index WHERE ((Level=20016) AND (Path IN ('metrics', 'list', ...)) AND (Date='1970-02-12') GROUP BY Path 
SELECT Path, groupArray(Time), groupArray(Value), grouypArray(Timestamp) FROM graphite.graphite PREWHERE Date >='2020-10-07' AND Date <= '2020-10-08' WHERE (Path IN ('metrics', 'list', ...) AND (Time >= 123456789 AND Time <= 987654321) GROUP BY Path FORMAT RowBinary
```

(5)
```bash
# A lot of
SELECT Path FROM graphite.graphite_index WHERE ((Level=20016) AND (Path IN ('metrics', 'list', ...)) AND (Date='1970-02-12') GROUP BY Path 
SELECT Path, groupArray(Time), groupArray(Value), grouypArray(Timestamp) FROM graphite.graphite PREWHERE Date >='2020-10-07' AND Date <= '2020-10-08' WHERE (Path IN ('metrics', 'list', ...) AND (Time >= 123456789 AND Time <= 987654321) GROUP BY Path FORMAT RowBinary
```

The SELECT queries in the logs are the same, so it looks like in my case the option has no effect.

**Comment by Felixoid** [comment]
Because necessary changes are in different branch =)
https://github.com/Felixoid/graphite-clickhouse/tree/clickhouse_aggregation

**Comment by mrodikov** [comment]
With a clickhouse_aggregation branch, I've got LA >90, memory 22/24G, then after 10+ minutes, carbonapi process was killed by OOM.

The experiment failed successfully I guess

**Comment by Felixoid** [comment]
Could you, maybe, ping me in telegram? Let's profile the applications and see some flamegraphs, maybe.

The username is the same there

**Comment by Civil** [comment]
Also as it's also happens on gch side, it would be great to see if you can send same query that causes huge memory consmuption with curl and dump the result in file and then try to interpret it with `protoc` CLI tools. That might have a clue what triggers increased memory consumption there.

p.s. if you are willing to talk over telegram, it would be great if you can create some sort of channel or something, where we can discuss that together.

**Comment by mrodikov** [comment]
In the end, it helped to set ``` maxBatchSize``` to a zero globally, not just for a specific backend.

For graphite-clickhouse the ```daily-index``` option was set to ```true```, the ```internal-aggregation``` option was set to ```true``` and binary was built from the latest ```master```.

Query execution time is 1-1.5 seconds now.

The issue can be closed I guess.
